adapters/aidem/aidem.go

```diff
@@ -0,0 +1,130 @@
+	var errs []error
+
+	reqJson, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJson,
+		Headers: headers,
+	}}, errs
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponse()
+
+	for _, seatBid := range bidResp.SeatBid {
+		for i := range seatBid.Bid {
+			bidType, err := getMediaTypeForImp(seatBid.Bid[i], internalRequest.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				b := &adapters.TypedBid{
+					Bid:     &seatBid.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+}
+
+// Builder builds a new instance of the AIDEM adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	return &adapter{
+		endpoint: config.Endpoint,
+	}, nil
+}
+
+func getMediaTypeForImp(bid openrtb2.Bid, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	var bidExt aidemResponseBidExt
+	err := json.Unmarshal(bid.Ext, &bidExt)
+	if err == nil {
+		switch bidExt.AdCodeType {
+		case "banner":
+			return openrtb_ext.BidTypeBanner, nil
+		case "native":
+			return openrtb_ext.BidTypeNative, nil
+		case "video":
+			return openrtb_ext.BidTypeVideo, nil
+		}
+	}
+
+	var mediaType openrtb_ext.BidType
+	var typeCnt = 0
+	for _, imp := range imps {
+		if imp.ID == bid.ImpID {
+			if imp.Banner != nil {
+				typeCnt += 1
+				mediaType = openrtb_ext.BidTypeBanner
+			}
+			if imp.Native != nil {
+				typeCnt += 1
+				mediaType = openrtb_ext.BidTypeNative
+			}
+			if imp.Video != nil {
+				typeCnt += 1
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+		}
+	}
+	if typeCnt == 1 {
+		return mediaType, nil
+	}
+	return mediaType, fmt.Errorf("unable to fetch mediaType in multi-format: %s", bid.ImpID)
+}
```

Comment:
Warning: We recommend setting media type based on adapter response. Use `bidresponse.seatbid.bid[i].ext.prebid.type` in the response from adapter to identify the mediatype of the imp.

Reply:
Thank you Sonali.
Could you share a code example from an existing adapter than can point us on how to achieve this? Thanks.

Reply:
@gsollazzo Slight change here. Please use the [MType](https://github.com/prebid/openrtb/blob/53ac1b9ca7ddf860988ff69bf5018b32ced1fbc5/openrtb2/bid.go#L334) field in the response to accurately determine the media type for the impression. This is a new field introduced in the openRTB spec rather than using `ext.prebid.type`. Here [an example](https://github.com/prebid/prebid-server/blob/6b616196605030f568dae4a727d1a658de929417/adapters/ix/ix.go#L246-L247) you can refer

Reply:
Hello @Sonali-More-Xandr, thank you.
It is my understanding that MType has been introduced in oRTB v2.6, therefore adding it to the adapter's protocol would imply a deviation from oRTB v2.5 (currently used by most adapters in this repo, and from my understanding the actual oRTB version marshalled by prebid-core).

Taking in account your [previous review,](https://github.com/prebid/prebid-server/pull/2824/files/4febd5afd3df6a82797b5eed1e0d6bd96d0f1982#diff-1bbf2c470ff681531013e56984e6fd000d5a68d8a1ab7324f6b070cbd30055b2) the only adapter I see using `if adapters.IsResponseStatusCodeNoContent(response) { 
   return nil, nil 
 }` is `appnexus`, but appnexus itself doesn't leverage MType, and relies instead of custom signalling via `appnexusBidExtAppnexus`: I'm not sure I understand the oRTB standard we need to conform to.

Is my understanding wrong? Shall we use native oRTB v2.6 directly?

I really like MType (our PrebidJS adapter rely on a similar mechanism), but before refactoring the adapter I would like to have clear requirements.

Thanks!

Reply:
Confirmed with the team that Prebid incorporates openRTB2.6. I agree that most of the adapters are not yet adhering to openRTB2.6. However, since this is a new adapter and we prefer to adhere to the openRTB2.6 specification, so requesting modification to utilize `MType`. 

Additionally, In the future, we plan to eliminate the requirement of setting BidType in `openrtb2.BidResponse.Bids.BidType` and instead directly employ `MType` to determine the bid type. I hope this clarification is helpful.

Reply:
Thank you, will proceed with v2.6 implementation.

Reply:
@Sonali-More-Xandr the adapter has been upgrade to oRTB v2.6 + MType + utils function.